ipc/openWindow.ts

```diff
@@ -12,8 +12,26 @@ const channel = {
     launcher: 'open-app-launcher', // It would be nice to call this `open:launcher` but we have to stick to the current name, because that is used by supported apps.
 };
 
+export const isOpenAppOptionsDeviceSN = (
+    device: OpenAppOptionsDevice
+): device is OpenAppOptionsDeviceSN =>
+    (device as OpenAppOptionsDeviceSN).serialNumber !== undefined;
+
+export const isOpenAppOptionsDevicePort = (
+    device: OpenAppOptionsDevice
+): device is OpenAppOptionsDevicePort =>
+    (device as OpenAppOptionsDevicePort).serialPortPath !== undefined;
+
+export type OpenAppOptionsDeviceSN = { serialNumber: string };
+
+export type OpenAppOptionsDevicePort = { serialPortPath: string };
+
+export type OpenAppOptionsDevice =
+    | OpenAppOptionsDeviceSN
+    | OpenAppOptionsDevicePort;
```

Comment:
The example below will ultimately do the same thing using Generic :) 

```typescript
interface OpenAppOptions {
    device?: 
        | { serialNumber: string; }
        | { serialPortPath: string; };
}

const getOpenAppOptions = <Device extends OpenAppOptions>(device: Device) => {
    return device;
}
```

Reply:
I made some test cases that illustrates how the type checking will work, and I think it's similar to how you wrote them. I'm a bit puzzled if we want to avoid the cases where you provide both `serialNumber` and `serialPortPath` or if we want to accept them?

```typescript
// ✅
const value1 = getOpenAppOptions({ device: { serialNumber: '0909090909' } });
type expected1 = { device: { serialNumber: string}};
type test1 = Expect<Equal<typeof value1, expected1>>

// ✅
const value2 = getOpenAppOptions({ device: { serialPortPath: '/dev/ttyACM0' } });
type expected2 = { device: { serialPortPath: string}};
type test2 = Expect<Equal<typeof value2, expected2>>

// ✅
const value3 = getOpenAppOptions({ device: undefined });
type expected3 = { device: undefined};
type test3 = Expect<Equal<typeof value3, expected3>>

// ✅
const value4 = getOpenAppOptions({});
type expected4 = {};
type test4 = Expect<Equal<typeof value4, expected4>>


// Does not work, because it expects device.
// @ts-expect-error ==> ❌
const invalid1 = getOpenAppOptions();

// This is actually valid: ✅
const invalid2 = getOpenAppOptions( { device: { serialPortPath: '0909090909', serialNumber: '/dev/ttyACM0' } } );
```

Reply:
Use case I have can be found here 
https://github.com/NordicSemiconductor/pc-nrfconnect-launcher/blob/fix/auto-connect-single-instance/src/main/windows.ts#L126